File: gateway/auth/go/sigs.go
package macaroon

import (
	"crypto/ed25519"
	"crypto/sha256"
	"fmt"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
)

// ─── Ed25519 ──────────────────────────────────────────────────────────

// Ed25519Sign signs msg with a 32-byte Ed25519 seed (the standard
// "private key" representation used in the TS sibling and in
// fixtures). Returns the 64-byte signature. Ed25519 is deterministic
// — same seed + same msg → same signature, byte-for-byte, across
// implementations.
func Ed25519Sign(seed, msg []byte) ([]byte, error) {
	if len(seed) != ed25519.SeedSize {
		return nil, fmt.Errorf("ed25519: seed must be %d bytes, got %d",
			ed25519.SeedSize, len(seed))
	}
	priv := ed25519.NewKeyFromSeed(seed)
	return ed25519.Sign(priv, msg), nil
}

// Ed25519Verify checks a 64-byte Ed25519 signature over msg against a
// 32-byte raw pubkey. Ed25519 signs the raw message (no prehash) per
// RFC 8032.
func Ed25519Verify(pubKey, msg, sig []byte) bool {
	if len(pubKey) != ed25519.PublicKeySize || len(sig) != ed25519.SignatureSize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pubKey), msg, sig)
}

// Ed25519PublicKey derives the 32-byte raw public key from a 32-byte
// seed. Useful for issuer key bootstrapping and tests.
func Ed25519PublicKey(seed []byte) ([]byte, error) {
	if len(seed) != ed25519.SeedSize {
		return nil, fmt.Errorf("ed25519: seed must be %d bytes, got %d",
			ed25519.SeedSize, len(seed))
	}
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	out := make([]byte, len(pub))
	copy(out, pub)
	return out, nil
}

// ─── ECDSA-secp256k1-SHA256 ───────────────────────────────────────────

// EcdsaSecp256k1Sign produces a 64-byte compact (r||s) ECDSA signature
// over secp256k1, SHA-256 prehashed, RFC 6979 deterministic, BIP 62
// low-s normalized. Cross-language byte-equivalence with TS sibling
// (which uses @noble/curves with the same defaults) depends on all
// three properties — don't substitute a randomized signer here.
//
// privKey is the 32-byte secp256k1 private scalar.
func EcdsaSecp256k1Sign(privKey, msg []byte) ([]byte, error) {
	if len(privKey) != 32 {
		return nil, fmt.Errorf("secp256k1: private key must be 32 bytes, got %d", len(privKey))
	}
	priv := secp256k1.PrivKeyFromBytes(privKey)
	digest := sha256.Sum256(msg)
	sig := ecdsa.Sign(priv, digest[:]) // RFC 6979 + low-s by construction
	r := sig.R()
	s := sig.S()
	out := make([]byte, 64)
	var rBuf, sBuf [32]byte
	r.PutBytes(&rBuf)
	s.PutBytes(&sBuf)
	copy(out[:32], rBuf[:])
	copy(out[32:], sBuf[:])
	return out, nil
}

// EcdsaSecp256k1Verify checks a 64-byte compact (r||s) ECDSA signature
// over secp256k1 with SHA-256 prehash. The signature MUST be low-s
// (BIP 62) — high-s signatures are rejected to keep "same logical
// signature → same bytes" enforceable.
//
// pubKey is the 33-byte compressed secp256k1 public key.
func EcdsaSecp256k1Verify(pubKey, msg, sig []byte) bool {
	if len(sig) != 64 {
		return false
	}
	pk, err := secp256k1.ParsePubKey(pubKey)
	if err != nil {
		return false
	}
	r := new(secp256k1.ModNScalar)
	if overflow := r.SetByteSlice(sig[:32]); overflow {
		return false
	}
	s := new(secp256k1.ModNScalar)
	if overflow := s.SetByteSlice(sig[32:]); overflow {
		return false
	}
	if r.IsZero() || s.IsZero() {
		return false
	}
	// Enforce low-s: reject signatures where s > n/2.
	if s.IsOverHalfOrder() {
		return false
	}
	digest := sha256.Sum256(msg)
	parsed := ecdsa.NewSignature(r, s)
	return parsed.Verify(digest[:], pk)
}

// EcdsaSecp256k1PublicKey derives the 33-byte compressed secp256k1
// public key from a 32-byte private scalar. Useful for issuer key
// bootstrapping and tests.
func EcdsaSecp256k1PublicKey(privKey []byte) ([]byte, error) {
	if len(privKey) != 32 {
		return nil, fmt.Errorf("secp256k1: private key must be 32 bytes, got %d", len(privKey))
	}
	priv := secp256k1.PrivKeyFromBytes(privKey)
	return priv.PubKey().SerializeCompressed(), nil
}

File: gateway/auth/ts/src/sigs.ts
/**
 * Signing + verifying primitives for the two curves phase 4 uses:
 * Ed25519 (user keys) and ECDSA-secp256k1-SHA256 (org keys, single-key
 * and multisig members).
 *
 * All inputs and outputs are raw `Uint8Array`s. Hex conversion happens
 * at the boundary in `sign.ts` / `verify.ts`.
 */

import { secp256k1 } from "@noble/curves/secp256k1";
import { ed25519 } from "@noble/curves/ed25519";
import { sha256 } from "@noble/hashes/sha2";

// ─── Ed25519 ──────────────────────────────────────────────────────────

/**
 * Sign raw bytes with Ed25519. No prehash — Ed25519 signs the message
 * directly (per RFC 8032).
 *
 * @param privKey 32-byte private key
 * @param msg     bytes to sign
 * @returns       64-byte signature
 */
export function ed25519Sign(privKey: Uint8Array, msg: Uint8Array): Uint8Array {
  return ed25519.sign(msg, privKey);
}

/**
 * @param pubKey 32-byte raw public key
 * @param msg    bytes that were signed
 * @param sig    64-byte signature
 */
export function ed25519Verify(
  pubKey: Uint8Array,
  msg: Uint8Array,
  sig: Uint8Array,
): boolean {
  try {
    return ed25519.verify(sig, msg, pubKey);
  } catch {
    return false;
  }
}

export function ed25519PublicKey(privKey: Uint8Array): Uint8Array {
  return ed25519.getPublicKey(privKey);
}

// ─── ECDSA-secp256k1-SHA256 ───────────────────────────────────────────

/**
 * Sign with ECDSA over secp256k1. We SHA-256 the message first
 * (phase 4 requires the prehash), then produce a compact `r||s`
 * signature with low-s normalization (BIP 62) so two signers
 * producing "the same" signature produce byte-identical output.
 *
 * @param privKey 32-byte private key
 * @param msg     bytes to sign (will be SHA-256'd internally)
 * @returns       64-byte compact signature (r||s), low-s normalized
 */
export function ecdsaSign(privKey: Uint8Array, msg: Uint8Array): Uint8Array {
  const digest = sha256(msg);
  // @noble/curves' sign() is low-s by default and returns a Signature
  // object; toCompactRawBytes() emits the 64-byte r||s form.
  const sig = secp256k1.sign(digest, privKey, { lowS: true });
  return sig.toCompactRawBytes();
}

/**
 * @param pubKey 33-byte compressed public key
 * @param msg    bytes that were signed (will be SHA-256'd internally)
 * @param sig    64-byte compact signature (r||s)
 * @returns      true iff signature is valid AND low-s normalized
 */
export function ecdsaVerify(
  pubKey: Uint8Array,
  msg: Uint8Array,
  sig: Uint8Array,
): boolean {
  try {
    const digest = sha256(msg);
    return secp256k1.verify(sig, digest, pubKey, { lowS: true });
  } catch {
    return false;
  }
}

/**
 * Compressed (33-byte) pubkey from a 32-byte private key.
 */
export function ecdsaPublicKey(privKey: Uint8Array): Uint8Array {
  return secp256k1.getPublicKey(privKey, true);
}
